File: hw/fib.c
#include <stdio.h>
#include <stdlib.h> 
#include <stdint.h>
#include <stdbool.h>

#define MAX 50000

typedef uint64_t ull;

static ull fib_table[MAX];

ull fib_iterative(int n) {
    uint64_t fibTotal = 0, fib1 = 0, fib2 = 1;
    for (int i = 0; i < n; i++) {
        fibTotal = fib1 + fib2;
        fib1 = fib2;
        fib2 = fibTotal;
    }
    return fib1;
}

ull fib_recursive(int n) {
    if (n == 0 || n == 1) {
        return n;
    } else {
        return fib_recursive(n-1) + fib_recursive(n-2);
    }
}

ull fib_dynamic(int n) {
    if (n == 0 || n == 1) {
        return n;
    }
    fib_lst[0] = 0;
    fib_lst[1] = 1;
    for (int i = 2; i <= n; i++) {
        fib_lst[i] = fib_lst[i - 1] + fib_table[i - 2];
    }
    ull answer = fib_lst[n];
    return answer;
}

int getFib(int n, int version) {
    int temp;
    if (version == 0) {
        temp = fib_iterative(n);
    } else if (version == 1) {
        temp = fib_recursive(n);
    } else {
        for (int i = 0; i < MAX; i++) {
            fib_table[i] = 0;
        }
        fib_table[1] = 1;
        temp = fib_dynamic(n);
    }
    return temp;
}

int* print_fib_series(int n, bool print, int version) {
    int* fibArr = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        int temp = getFib(i, version);
        fibArr[i] = temp;
        if (print) {
            printf("%d\t", temp);
        }
    }
    if (print) {
        printf("\n");
    }
    free(fibArr);
    return fibArr;
}

void help() {
printf("./fibonacci.out N [Type] [Print Level]\n");
printf("\tN is the number in the Fibonacci series to calculate, required.\n");
printf("\t[Type] is either 2 for dynamic programming version, 1 for recursive version, 0 for iterative version, defaults to 0.\n");
printf("\t[Print Level] is 0 for no print/speed compare only, 1 for print nth value only, 2 for print first N values, defaults to 0.\n");
}


int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Need 2 or more arguments.\n");
        help();
        return 1;
    }

    const int n = atoi(argv[1]);
    int version = 0;
    int printLevel = 0;

    if (argc > 2) {
        version = atoi(argv[2]);
    }

    if (argc > 3) {
        printLevel = atoi(argv[3]);
    }
    bool print = false;
    if (printLevel > 1) {
        print = true;
    }
    int* fib_series = print_fib_series(n, print, version);
    if (printLevel == 1) {
        printf("%d\n", getFib(n, version));
    }

    return 0;
}
    

File: hw/fib.py
from enum import Enum
from typing import List
from functools import lru_cache
import click
import sys

sys.setrecursionlimit(100000)


class FibType(Enum):
    DP = 2
    RECURSIVE = 1
    ITERATIVE = 0


def fib_iterative(n: int) -> int:
    """
        Calculates the nth Fibonacci number using an iterative approach.

        Args:
            n: An integer specifying which Fibonacci number to calculate.

        Returns:
            An integer representing the nth Fibonacci number.
        """
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        first = 0
        second = 1
        for i in range(2, n+1):
            next_num = first + second
            first = second
            second = next_num
        return second


def fib_recursive(n: int) -> list[int]:
    """
        Calculates the nth Fibonacci number using a recursive approach.

        Args:
            n: An integer specifying which Fibonacci number to calculate.

        Returns:
            A list of integers representing the first n Fibonacci numbers.
        """
    if n <= 0:
        return [0]
    elif n == 1:
        return [0, 1]
    else:
        fib_list = fib_recursive(n - 1)
        fib_list.append(fib_list[-1] + fib_list[-2])
        return fib_list


@lru_cache(maxsize=None)
def fib_dp(n: int) -> list[int]:
    """
        Calculates the nth Fibonacci number using dynamic programming with memoization.

        Args:
            n: An integer specifying which Fibonacci number to calculate.

        Returns:
            A list of integers representing the first n Fibonacci numbers.
        """
    fib_lst = [-1] * (n + 1)
    fib_lst[0] = 0
    fib_lst[1] = 1
    for i in range(2, n + 1):
        fib_lst[i] = fib_lst[i - 1] + fib_lst[i - 2]
    return fib_lst


def get_fib(n: int, version: FibType) -> List[int]:
    """
        Returns the first n Fibonacci numbers using the specified calculation method.

        Args:
            n: An integer specifying how many Fibonacci numbers to calculate.
            version: A member of the FibType enum specifying which calculation method to use.

        Returns:
            A list of integers representing the first n Fibonacci numbers.
        """
    if version == FibType.ITERATIVE:
        return [fib_iterative(i) for i in range(n)]
    elif version == FibType.RECURSIVE:
        return fib_recursive(n)
    elif version == FibType.DP:
        return fib_dp(n)[:n]


def print_fib(n: int, print_it: bool = False, version: FibType = FibType.ITERATIVE) -> List[int]:
    """
        Computes and prints the first n numbers in the Fibonacci sequence using the specified method.

        Args:
            n (int): The number of Fibonacci numbers to compute.
            print_it (bool, optional): Whether to print each number as it is computed. Defaults to False.
            version (FibType, optional): The method to use for computing the Fibonacci numbers. Defaults to FibType.ITERATIVE.

        Returns:
            List[int]: A list of the first n Fibonacci numbers.
    """
    if version == FibType.RECURSIVE:
        return fib_recursive(n)
    if version == FibType.DP:
        return fib_dp(n)
    for i in range(n):
        fib_lst = [fib_iterative(i) for i in range(n)]
        if print_it:
            print(fib_lst[-1], end="")
    if print_it:
        print()
    return fib_lst


@click.command()
@click.argument("n", type=click.IntRange(min=0, max=50000, clamp=True))
@click.option("--algo", type=click.Choice(['recursive', 'dp', 'iterative'], case_sensitive=False), default='iterative')
@click.option("--print-type", type=click.Choice(['all', 'none', 'single'], case_sensitive=False), default='none')
def main(n: int, algo: str, print_type: str):
    print_it = print_type == 'all'
    t = FibType.ITERATIVE
    if algo == 'recursive':
        t = FibType.RECURSIVE
    elif algo == 'dp':
        t = FibType.DP
    fib_nums = get_fib(n, t)
    if print_type == 'single':
        print(fib_nums[-1])


if __name__ == "__main__":
    main()
